hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+        metrics.updateMetadataMetrics(VALIDATE_FILES_STR, validateTime, numLookupsFilesInPartition);
+      }
+    }
+
+    if (context != null) {
+      long durationInMs = metrics.getDurationInMs(context.stop()) - validateTime;
+      metrics.updateMetadataMetrics(LOOKUP_FILES_STR, durationInMs, numLookupsFilesInPartition);
+    }
+
+    return statuses;
+  }
+
+  /**
+   * Open readers to the base and log files.
+   */
+  private synchronized void openBaseAndLogFiles() throws IOException {
+    if (logRecordScanner != null) {
+      // Already opened
+      return;
+    }
+
+    final Timer.Context context = this.metrics.getMetadataCtx(SCAN_STR);
+
+    // Find the latest file slice
+    HoodieTimeline timeline = metaClient.reloadActiveTimeline();
+    HoodieTableFileSystemView fsView = new HoodieTableFileSystemView(metaClient, metaClient.getActiveTimeline());
+    List<FileSlice> latestSlices = fsView.getLatestFileSlices(METADATA_PARTITION_NAME).collect(Collectors.toList());
+    ValidationUtils.checkArgument(latestSlices.size() == 1);
+
+    // If the base file is present then create a reader
+    Option<HoodieBaseFile> basefile = latestSlices.get(0).getBaseFile();
+    if (basefile.isPresent()) {
+      String basefilePath = basefile.get().getPath();
+      basefileReader = HoodieFileReaderFactory.getFileReader(hadoopConf, new Path(basefilePath));
+
+      LOG.info("Opened metadata base file from " + basefilePath + " at instant " + basefile.get().getCommitTime());
+    }
+
+    // Open the log record scanner using the log files from the latest file slice
+    List<String> logFilePaths = latestSlices.get(0).getLogFiles().map(o -> o.getPath().toString())
+        .collect(Collectors.toList());
+    Option<HoodieInstant> lastInstant = timeline.filterCompletedInstants().lastInstant();
+    String latestTimestamp = lastInstant.isPresent() ? lastInstant.get().getTimestamp() : "";
+
+    logRecordScanner =
+        new HoodieMetadataMergedLogRecordScanner(FSUtils.getFs(basePath, hadoopConf), metadataBasePath,
+            logFilePaths, schema, latestTimestamp, maxMemorySizeInBytes, bufferSize,
+            config.getSpillableMapBasePath(), null);
+
+    LOG.info("Opened metadata log files from " + logFilePaths + " at instant " + lastInstant);
+
+    ++numScans;
+
+    if (context != null) {
+      long durationInMs = metrics.getDurationInMs(context.stop());
+      metrics.updateMetadataMetrics(SCAN_STR, durationInMs, numScans);
+    }
+  }
+
+  /**
+   * Retrieve the merged {@code HoodieRecord} mapped to the given key.
+   *
+   * @param key The key of the record
+   */
+  private Option<HoodieRecord<HoodieMetadataPayload>> getMergedRecordByKey(String key) throws IOException {
+    openBaseAndLogFiles();
+
+    // Retrieve record from base file
+    HoodieRecord<HoodieMetadataPayload> hoodieRecord = null;
+    if (basefileReader != null) {
+      Option<GenericRecord> baseRecord = basefileReader.getRecordByKey(key);
```

Comment:
note to self: typical use-case is to list all partition paths and then for each pp, find all the file slices etc. so probably merging the base and log in each such call is okay? with inlined hFIle logs in log. this can be much faster.